File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/MenuActivity.java
package com.productions.gizzmoo.pokemonpuzzleleague;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;

public class MenuActivity extends AppCompatActivity {

    private final String currentFrameKey = "CURRENT_FRAME_KEY";
    private final String isPanningRightKey = "IS_PANNING_RIGHT_KEY";

    private final int MAX_FPS = 30;
    private final int FRAME_PERIOD = 1000 / MAX_FPS;
    private final int NUM_OF_SECS_TO_PAN_VIEW = 10;

    private ImageViewPanning mPanningImage;
    private PanningLoop mPanningLoop;
    private int mCurrentFrame;
    private boolean mIsPanningRight;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        Button mStartButton = (Button) findViewById(R.id.startButton);
        mStartButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(MenuActivity.this, TimeZoneActivity.class);
                startActivity(intent);
                }
        });

        mPanningImage = (ImageViewPanning) findViewById(R.id.backgroundPanning);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        if (savedInstanceState != null) {
            mCurrentFrame = savedInstanceState.getInt(currentFrameKey);
            mIsPanningRight = savedInstanceState.getBoolean(isPanningRightKey);
        } else {
            mCurrentFrame = NUM_OF_SECS_TO_PAN_VIEW * MAX_FPS / 2;
            mIsPanningRight = true;
        }
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            getWindow().getDecorView().setSystemUiVisibility(
                    View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                            | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION | View.SYSTEM_UI_FLAG_FULLSCREEN | View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_setting:
                Intent intent = new Intent(MenuActivity.this, SettingsActivity.class);
                startActivity(intent);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onStart() {
        super.onStart();

        mPanningLoop = new PanningLoop();
        mPanningLoop.execute();
    }

    @Override
    public void onStop() {
        super.onStop();

        if (mPanningLoop != null) {
            mPanningLoop.cancel(true);
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(currentFrameKey, mPanningImage.getCurrentImageFrame());
        outState.putBoolean(isPanningRightKey, mPanningImage.isPanningRight());
    }

    private class PanningLoop extends AsyncTask<Void, Void, Void> {

        @Override
        protected Void doInBackground(Void... voids) {
            while (!mPanningLoop.isCancelled()) {
                try {
                    Thread.sleep(FRAME_PERIOD);
                } catch (InterruptedException e) {}

                publishProgress();
            }

            return null;
        }

        @Override
        protected void onProgressUpdate(Void... values) {
            mPanningImage.moveFrame();
            mPanningImage.invalidate();
        }

        @Override
        protected void onPreExecute() {
            mPanningImage.framesToLookAtView(NUM_OF_SECS_TO_PAN_VIEW * MAX_FPS);
            mPanningImage.setCurrentFrame(mCurrentFrame);
            mPanningImage.setDirectionRight(mIsPanningRight);
        }
    }
}

File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/MenuActivity.kt
package com.productions.gizzmoo.pokemonpuzzleleague

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.widget.Button
import com.productions.gizzmoo.pokemonpuzzleleague.PanningLoop.Companion.MAX_FPS
import com.productions.gizzmoo.pokemonpuzzleleague.PanningLoop.Companion.NUM_OF_SECS_TO_PAN_VIEW
import com.productions.gizzmoo.pokemonpuzzleleague.puzzlegame.puzzleacademygame.PuzzleAcademySelectionActivity
import com.productions.gizzmoo.pokemonpuzzleleague.puzzlegame.marathongame.MarathonActivity
import com.productions.gizzmoo.pokemonpuzzleleague.puzzlegame.spaservicegame.SpaServiceGameActivity
import com.productions.gizzmoo.pokemonpuzzleleague.settings.SettingsActivity

class MenuActivity : AppCompatActivity(), PanningLoopListener {

    private val currentFrameKey = "CURRENT_FRAME_KEY"
    private val directionKey = "DIRECTION_KEY"

    private lateinit var panningImage: ImageViewPanning
    private var panningLoop: PanningLoop? = null
    private var currentFrame: Int = 0
    private var direction: Direction = Direction.None

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)
        val marathonButton = findViewById<Button>(R.id.marathonButton)
        marathonButton.setOnClickListener {
            val intent = Intent(this@MenuActivity, MarathonActivity::class.java)
            startActivity(intent)
            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
        }

        val puzzleAcademyButton = findViewById<Button>(R.id.puzzleAcademy)
        puzzleAcademyButton.setOnClickListener {
            val intent = Intent(this@MenuActivity, PuzzleAcademySelectionActivity::class.java)
            startActivity(intent)
            overridePendingTransition(R.anim.slide_out_left, R.anim.slide_in_right)
        }

        val spaServiceButton = findViewById<Button>(R.id.spaService)
        spaServiceButton.setOnClickListener {
            val intent = Intent(this@MenuActivity, SpaServiceGameActivity::class.java)
            startActivity(intent)
            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
        }

        panningImage = findViewById(R.id.backgroundPanning)

        if (savedInstanceState != null) {
            currentFrame = savedInstanceState.getInt(currentFrameKey)
            direction = savedInstanceState.getSerializable(directionKey) as Direction
        } else {
            currentFrame = NUM_OF_SECS_TO_PAN_VIEW * MAX_FPS / 2
            direction = Direction.Right
        }

        panningImage.framesNeededForView = NUM_OF_SECS_TO_PAN_VIEW * MAX_FPS
        panningImage.setCurrentFrame(currentFrame)
        panningImage.direction = direction
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
        when (item.itemId) {
            R.id.action_setting -> {
                val intent = Intent(this@MenuActivity, SettingsActivity::class.java)
                startActivity(intent)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }

    override fun onStart() {
        super.onStart()
        panningLoop = PanningLoop()
        panningLoop?.listener = this
        panningLoop?.execute()
    }

    override fun onStop() {
        super.onStop()
        panningLoop?.cancel(true)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(currentFrameKey, panningImage.currentFrame)
        outState.putSerializable(directionKey, panningImage.direction)
    }

    override fun updateImageView() {
        panningImage.moveFrame()
        panningImage.invalidate()
    }
}
